[WEB-211] Account profile redesign

314numberpi opened the pull request.

EladKarni committed:
Initial Draft of Overview Component Done

EladKarni committed:
another redesign of profile header

314numberpi committed:
Merge pull request #160 from EladKarni/redesign/WEB-211/Profile-Overview-component-Creation

Redesign/web 211/profile overview component creation

314numberpi committed:
Fix(ProfileHeader-components): Fixed avatar

314numberpi committed:
Refactor(ProfileHeader-scss): Removed redundant styles, fixed issues

314numberpi committed:
Merge branch 'develop' into feature/WEB-211/profile-account

314numberpi committed:
feat(userNicknames): added nick count

314numberpi committed:
refactor(userNicknames): more compact style

314numberpi committed:
refactor(userNicknames): add add nickname toggle

314numberpi committed:
Style(profile-header): fixed indentation

314numberpi committed:
Refactor(ProfileHeader): changed date format

314numberpi committed:
Refactor(UserDetailsPanel): Removed UserDetailsPanel

314numberpi committed:
refactor(UserNicknamesPanel): added check for max nicks reached

314numberpi committed:
refactor(UserDecalPanel): add decal toggle button[WIP]

UncleClapton commented on client/helpers/getDisplayRat.js:
```diff
@@ -0,0 +1,20 @@
+const getDisplayRat = (user, ratState) => {
```
Not sure how this got added back again. File should be removed.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
+      <div className="decal-code" key={`code${index}`}>
+        <button
+          key={`toggle${index}`}
+          aria-label="toggle-decal"
```
aria-label should be human readable explanatory text, as it is read by screen readers.
```suggestion
          aria-label="Show decal code"
```

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
+    <>
+      <div className="decal-code" key={`code${index}`}>
+        <button
+          key={`toggle${index}`}
```
unnecessary key being set here. only the root element returned in a map needs the key.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
   }
 
+  _renderDecalCode = (decal, index) => (
+    <>
```
Avoid using fragments in maps when possible. This should probably be a wrapper `<div />` with an identifying className like `decal-row`

314numberpi replied:
The problem with that is it messes up the formatting, the grid requires all direct child elements to function correctly (sub-grids, when?), otherwise the date column would either have variable widths for each row, or a fixed-width, not based on the content.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
 
+  _renderDecalCode = (decal, index) => (
+    <>
+      <div className="decal-code" key={`code${index}`}>
```
You should never use `index` in a key attribute. 
Since we are mapping out a decal resource, we should use it's id

```suggestion
      <div className="decal-code" key={decal.id}>
```

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
+      return null
+    }
+
+    return (
```
no need for the parenthesis here. you can just do
```javascript
return decals.map(this._renderDecalCode)
```

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
     return null
   }
 
+  _renderDecalCode = (decal, index) => (
```
render methods should be "public" (without the `_`), and placed directly above the main render method.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
+          onClick={this._handleToggle}
+          // title={maxNicksReached ? 'You\'ve used all your nicknames' : 'Add new nickname'}
+          type="button">
+          <FontAwesomeIcon icon={this.state.formOpen[index] ? 'eye-slash' : 'eye'} fixedWidth />
```
again, avoid using index. Better to use the decal's id like we do with key attributes.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
+        </button>
+        {decal.attributes.code}
+      </div>
+      <div className="decal-claimed-at" key={`claimedAt${index}`}>{formatAsEliteDate(decal.attributes.claimedAt)}</div>
```
same suggestion on keys as above, however if we make the wrapper element a div then this can just be removed.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -108,26 +142,47 @@ class UserDetailsPanel extends Component {
+        <header>Decal</header>
+        <div className="panel-content">
+          {(checkingEligibility || redeeming) ? (<div className="loading">{loadingText}</div>) : (this._renderDecalCodes(decals))}
+          {(decals === []) ? (this._renderNoDataText()) : ('')}
```
Same as above, never compare to `[]`

314numberpi replied:
I don't know what you mean by this or why, or even how to do this otherwise?

Wait... length check perhaps...

UncleClapton replied:
wtf... the other one didn't get saved :/

314numberpi committed:
(getDisplayRat): Delete unused helper

314numberpi committed:
Fix(UserDecalPanel): Fix some silly mistakes.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,6 +90,37 @@ class UserDetailsPanel extends Component {
+    const { decals } = this.props
+    console.log(decals)
+
+    if (decals === []) {
```
Never compare to `[]` as it will always return false. using `[]` always creates a new instance of an object, meaning regardless of contents the two will never be equal. Since decals is always an array, it's safe to do a truthy check on `decals.length`.
```suggestion
    if (!decals.length) {
```

314numberpi committed:
refactor(UserDecalPanel): add toggle button functionality

314numberpi committed:
fix(UserDecalPanel): trigger button pointer events correctly

Disallow pointer events from triggering on the button's children, rather 
than the button itself

314numberpi committed:
refactor(UserDecalPanel): remove old and test code, and remove grid

314numberpi committed:
fix(UserDecalPanel): format no decals correctly

UncleClapton committed:
Merge branch 'develop' into feature/WEB-211/profile-account

314numberpi committed:
Refactor(UserNicknamesPanel): re-style addNicknameForm

UncleClapton commented on client/components/AddNicknameForm.js:
```diff
@@ -102,7 +122,11 @@ class AddNicknameForm extends Component {
   \***************************************************************************/
 
-  static mapDispatchToProps = ['addNickname']
+  static mapDispatchToProps = ['addNickname', 'addNicknamePass']
```
When `mapDispatchToProps` is given an array of strings, it's telling the `@connect` decorator to map the function of the same name found within the files of the `/client/store/actions` directory. I feel like this whole 'lets define a magic array of strings' thing has been confusing for a lot of people who has seen our website. might just remove it entirely soon once we move to the new namespaced functions thing I keep squeaking about.

Long story short, 'addNicknamePass' is not a valid store function, and there's no reason for it to be there.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -16,22 +16,21 @@ import Component from './Component'
   state = {
-    checkingEligibility: true,
-    redeeming: false,
+    ...initialState,
```
No reason for an `initialState` variable to be defined here as we don't need to clear state for any reason.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,25 +109,31 @@ class UserDetailsPanel extends Component {
+          type="button">
+          <FontAwesomeIcon icon={this.state.decalsVisible[decal.id] ? 'eye' : 'eye-slash'} fixedWidth />
+        </button>
+        {this.state.decalsVisible[decal.id] ? decal.attributes.code : `•••••-•••••-•••••-•••••-${decal.attributes.code.substring(24)}`}
```
Could we multiline long ternary expressions like this?

```suggestion
        {
          this.state.decalsVisible[decal.id] 
            ? decal.attributes.code 
            : `•••••-•••••-•••••-•••••-${decal.attributes.code.substring(24)}`
        }
```

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -54,13 +53,36 @@ class UserDetailsPanel extends Component {
 
-  static _renderClaimedAtRow (row) {
-    const rescue = row.original
+  _handleToggle = (event) => {
```
I'd name this something more descriptive like
```suggestion
  _handleDecalVisibilityToggle = (event) => {
```

UncleClapton commented on client/components/AddNicknameForm.js:
```diff
@@ -36,24 +39,28 @@ class AddNicknameForm extends Component {
 
     this.setState({ submitting: true })
 
     /* eslint-disable no-alert */
```
we can remove this eslint disable and enable lines now that we are no longer using a prompt to collect the password.

UncleClapton commented on client/components/ProfileHeader.js:
```diff
@@ -0,0 +1,120 @@
+
+
+
+// Component Constants
```
Could you do a round of clean up on all the files changed/introduced by this PR and remove empty section comments like this?

UncleClapton commented on client/components/ProfileHeader.js:
```diff
@@ -0,0 +1,120 @@
+
+
+
+  /***************************************************************************\
```
empty sections should not have headers like this.

UncleClapton commented on client/components/ProfileHeader.js:
```diff
@@ -0,0 +1,120 @@
+    Redux Properties
+  \***************************************************************************/
+
+  static mapDispatchToProps = ['getUser']
```
getUser is unused and not needed here. Further, if the user is already authenticated there is no need to ever call this from a component.

Could just entirely remove this line

UncleClapton commented on client/components/ProfileHeader.js:
```diff
@@ -0,0 +1,120 @@
+  /***************************************************************************\
+    Private Methods
+  \***************************************************************************/
+  _renderUserGroups = () => this.props.user.relationships.groups.data.map((item) => <li className={`badge ${item.id}`} key={item.id}>{item.id}</li>)
```
Please split this up a little. At least where it becomes JSX so it's not all forced onto one line.
Kind of like:
```suggestion
  _renderUserGroups = () => this.props.user.relationships.groups.data.map((item) => (
    <li className={`badge ${item.id}`} key={item.id}>
      {item.id}
    </li>
  ))
```

Though that might not be enough. If the user has no groups, data will most likely be null. You should probably run a null check before mapping.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -116,18 +144,16 @@ class UserDetailsPanel extends Component {
+        <header>Decal</header>
+        <div className="panel-content">
+          {(checkingEligibility || redeeming) ? (<div className="loading">{loadingText}</div>) : (this.renderDecalCodes(decals))}
+          {(decals.length) ? ('') : (this.renderNoDataText())}
```
I would personally write this like:
```suggestion
          {!decals.length && this.renderNoDataText()}
```

React won't render anything if a block evaluates to `false`, `null`, or `undefined`.

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -116,18 +144,16 @@ class UserDetailsPanel extends Component {
+      <div className="panel user-decals">
+        <header>Decal</header>
+        <div className="panel-content">
+          {(checkingEligibility || redeeming) ? (<div className="loading">{loadingText}</div>) : (this.renderDecalCodes(decals))}
```
Another long ternary to be multi-lined

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -139,31 +165,6 @@ class UserDetailsPanel extends Component {
     Getters
```
Another empty getter section

UncleClapton commented on client/components/UserNicknamesPanel.js:
```diff
@@ -4,55 +4,87 @@ import React from 'react'
 @connect
 class UserNicknamesPanel extends Component {
-  _deleteNickname = async (event) => {
+  _handleDeleteNickname = async (event) => {
```
could we restructure this to follow the template, and add in the section headers?

Class Properties (like state)
Private Methods (_ methods/functions)
Public Methods
Getters
Redux Mapping Properties
Prop Definitions

UncleClapton commented on client/components/UserNicknamesPanel.js:
```diff
@@ -4,55 +4,87 @@ import React from 'react'
+    formOpen: false,
+  }
+
+  _handleToggle = () => {
```
I'd rename this to something more descriptive like
```suggestion
  _handleFormVisibilityToggle = () => {
```

UncleClapton commented on client/components/UserNicknamesPanel.js:
```diff
@@ -4,55 +4,87 @@ import React from 'react'
     const {
       user,
     } = this.props
+    const maxNicksReached = (user.attributes.nicknames.length === MAXNICKS)
```
could we make this `>=` just for safety?

```suggestion
    const maxNicksReached = (user.attributes.nicknames.length >= MAXNICKS)
```

UncleClapton commented on client/components/UserNicknamesPanel.js:
```diff
@@ -4,55 +4,87 @@ import React from 'react'
-        <footer>
-          <AddNicknameForm />
-        </footer>
+        {/* (this.state.formOpen) && (
```
old garbage?

UncleClapton commented on client/helpers/formatTime.js:
```diff
@@ -14,10 +14,13 @@ const formatAsEliteDateTime = (timestamp) => moment(timestamp).add(ELITE_GAME_YE
 
 const formatAsEliteDate = (timestamp) => moment(timestamp).add(ELITE_GAME_YEAR_DESPARITY, 'years').format('DD MMM YYYY').toUpperCase()
 
+const formatAsEliteDateLong = (timestamp) => moment(timestamp).add(ELITE_GAME_YEAR_DESPARITY, 'years').format('DD MMMM YYYY')
```
yum more date formatters! 😋

UncleClapton commented on client/components/UserNicknamesPanel.js:
```diff
@@ -4,55 +4,87 @@ import React from 'react'
+
+  _handleToggle = () => {
+    this.setState((state) => ({
+      ...state,
```
no need to spread state like this normally. `setState` shallow merges the object you pass it. the only time we need to spread existing state is when we're trying to change something within an object stored within state, without losing the rest of that object.

This `setState` can just become
```javascript
this.setstate((state) => { formOpen: !state.formOpen })
```

UncleClapton commented on client/components/ProfileHeader.js:
```diff
@@ -0,0 +1,120 @@
+      userAvatar,
+    } = this.props
+
+    let { attributes } = this.props.user
```
These 2 lines could actually be
```suggestion
    const attributes = this.props.user.attributes || {}
```

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -87,25 +109,31 @@ class UserDetailsPanel extends Component {
     }
 
-    this.setState({ checkingEligibility: false })
+    return decals.map(this.renderDecalCode)
```
you could actually write this bit out as 
```suggestion
    return Boolean(decals.length) && decals.map(this.renderDecalCode)
```
 and remove that `if (!decals.length)` block above.

UncleClapton commented on client/helpers/formatTime.js:
```diff
@@ -14,10 +14,13 @@ const formatAsEliteDateTime = (timestamp) => moment(timestamp).add(ELITE_GAME_YE
 
 
 export {
   // eslint-disable-next-line import/prefer-default-export
```
Since we have more than one export from this file now, we can remove this config override.

UncleClapton reviewed (changes requested):
Here comes the big one 🤣 

Overall really good stuff. I'm loving how this is turning out. As we discussed lets hold off on the last two items for another ticket. I'm going to remove them both and put them into separate tickets.

314numberpi committed:
fix(formatTime-helper): remove eslint override

314numberpi committed:
style(ProfileHeader): Clean section headers, use clearer function names

314numberpi committed:
style(UserNicknamesPanel): Clean section headers, use clearer function names

314numberpi committed:
style(UserDecalPanel): clean section headers, use clearer function names

314numberpi committed:
Fix(AddNicknameForm): remove uneeded eslint override and redux map

UncleClapton commented on client/components/UserDecalPanel.js:
```diff
@@ -16,28 +16,11 @@ import Component from './Component'
-/***********************************************/
 @connect
 class UserDetailsPanel extends Component {
-  /***************************************************************************\
```
We still need state, it just doesn't need to be in an initialState variable 🤣

kenneaal commented:
These additional testing comments brought to you by Clapton.

kenneaal commented:
Who is entirely to blame!

314numberpi committed:
fix(UserDecalPanel): re-add state with previous initialState values

UncleClapton reviewed (approved):
LGTM!
